### PR TITLE
Improve plot_paired_histogram

### PR DESCRIPTION
**Pull request by JEM-Mosig**
**Proposed changes**:
- Rename `plot_histogram` to `plot_paired_histogram`, add tests, and fix an issue with a missing bar in the chart
- Closes https://github.com/RasaHQ/rasa/issues/7912

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- ~~updated the documentation~~
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by dakshvar22** [comment]
@JEM-Mosig Terribly sorry, I forgot about this. Somehow our slack reminder for pending reviews is not working and I completely missed this. I'll make sure to review this tomorrow.

**Comment by JEM-Mosig** [comment]
> @JEM-Mosig Terribly sorry, I forgot about this. Somehow our slack reminder for pending reviews is not working and I completely missed this. I'll make sure to review this tomorrow.

No worries.

File: rasa/utils/plotting.py [hunk 1]
```diff
@@ -117,103 +117,134 @@ def plot_confusion_matrix(
...
+    Returns:
+        The bins, values, ranges of either x-axis, and the range of the y-axis
     """
-    import matplotlib.pyplot as plt
-
-    plt.gcf().clear()
+    min_data_value = np.min(np.concatenate(histogram_data))
+    max_data_value = np.max(np.concatenate(histogram_data))
+    bin_width = (max_data_value - min_data_value) / num_bins
+    bins = [min_data_value + i * bin_width for i in range(num_bins + 2)]
```
Comment (dakshvar22) [inline rasa/utils/plotting.py:142]: Could you please add a comment why `+2` is needed?

File: rasa/utils/plotting.py [hunk 2]
```diff
@@ -117,103 +117,134 @@ def plot_confusion_matrix(
...
+    histograms = [
+        np.histogram(data, bins=bins, density=density)[0] for data in histogram_data
+    ]
+
+    y_padding = 0.5 * bin_width + y_pad_fraction * bin_width
+
+    if density:
```
Comment (dakshvar22) [inline rasa/utils/plotting.py:149]: Is this option being used anywhere in the code currently?

File: rasa/utils/plotting.py [hunk 3]
```diff
@@ -117,103 +117,134 @@ def plot_confusion_matrix(
...
+        # When we plot the PDF, let both x-axes run to the same value
+        # so it's easier to compare visually
+        v = max([(1.0 + x_pad_fraction) * max(histogram) for histogram in histograms])
+        x_ranges = [v, v]
+    else:
+        x_ranges = [(1.0 + x_pad_fraction) * max(histogram) for histogram in histograms]
```
Comment (dakshvar22) [inline rasa/utils/plotting.py:155]: Could you add a comment here describing how you are selecting the ranges. It's hard to guess without knowing the structure of `histograms`?

File: rasa/utils/plotting.py [hunk 4]
```diff
@@ -117,103 +117,134 @@ def plot_confusion_matrix(
...
-    )
-    min_value = min(
-        [min(hist_data[0], default=0), min(hist_data[1], default=0)], default=0
+    bin_of_first_non_zero_tally = min(
+        [(histogram != 0).argmax(axis=0) for histogram in histograms]
     )
 
-    bin_width = (max_value - min_value) / n_bins
-    bins = [min_value + (i * bin_width) for i in range(1, n_bins + 1)]
+    y_range = (bins[bin_of_first_non_zero_tally] - y_padding, bins[-2] + y_padding)
```
Comment (dakshvar22) [inline rasa/utils/plotting.py:161]: Same here please.

File: tests/utils/test_plotting.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import numpy as np
+import pytest
+
+
+@pytest.mark.parametrize(
+    "data, num_bins, expected_bins",
+    [
+        ([[1, 3, 8], [2, 3, 3]], 7, list(range(1, 9 + 1))),
+        ([[3, 8], [2, 3, 3]], 6, list(range(2, 9 + 1))),
+        ([[3, 7], [2, 3, 3]], 5, list(range(2, 8 + 1))),
```
Comment (dakshvar22) [inline tests/utils/test_plotting.py:12]: Is there a reason why the arguments inside `range` is specified as `9 + 1` and not `10`. If there is something subtle that you want to highlight with the offset `1`, then probably good to compliment it with a comment as well.

File: rasa/utils/plotting.py [hunk 6]
```diff
@@ -117,103 +117,134 @@ def plot_confusion_matrix(
         fig.savefig(output_file, bbox_inches="tight")
 
 
-def plot_histogram(
-    hist_data: List[List[float]], title: Text, output_file: Optional[Text] = None
-) -> None:
-    """Plots a side-by-side comparative histogram of the confidence distribution.
+def _paired_histogram_specification(
```
Comment (dakshvar22) [inline rasa/utils/plotting.py:120]: Can you rename the function to something like `_extract_paired_histogram_specification`. It tells that the function is doing something, i.e. `extract`.

File: tests/utils/test_plotting.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    assert np.all(histograms[0] == expected_histograms[0])
+    assert np.all(histograms[1] == expected_histograms[1])
+
+
+@pytest.mark.parametrize(
+    "data, num_bins, density, x_pad_fraction, expected_ranges",
+    [
+        ([[1, 3, 8], [2, 3, 3]], 100, False, 0.0, [1.0, 2.0]),
+        ([[1, 3, 8], [2, 3, 3, 3, 3]], 100, False, 0.0, [1.0, 4.0]),
+        ([[1, 3, 8], [2, 3, 3]], 7, True, 0.0, [2 / 3, 2 / 3]),
```
Comment (dakshvar22) [inline tests/utils/test_plotting.py:77]: `x_pad_fraction` is always `0` in these test cases. Can you add a couple non-zero test cases?

File: tests/utils/test_plotting.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        y_pad_fraction=0,
+    )
+    assert np.all(x_ranges == expected_ranges)
+
+
+@pytest.mark.parametrize(
+    "data, num_bins, y_pad_fraction, expected_range",
+    [
+        ([[1, 3, 8], [2, 3, 3]], 7, 0.0, [0.5, 8.5]),
+        ([[1, 3, 8], [2, 3, 3, 3, 3]], 7, 0.0, [0.5, 8.5]),
```
Comment (dakshvar22) [inline tests/utils/test_plotting.py:101]: non-zero `y_pad_fractions` missing in all test cases.

**Review by dakshvar22 — COMMENTED** [review]
Nice cleanup and testing! Added a few comments.

File: rasa/utils/plotting.py (hunk 2 above)
Comment (JEM-Mosig) [reply rasa/utils/plotting.py:149]: No, but I thought we might want to try it out. With densities you don't have the problem of different x-axis scales.

**Comment by JEM-Mosig** [comment]
@dakshvar22 Can you please review again?

File: rasa/utils/plotting.py [hunk 9]
```diff
@@ -167,30 +167,49 @@ def _paired_histogram_specification(
...
+        # The y_padding adds half a bin width, as we want the bars to be
+        # _centered_ on the bins. We take the next-to-last element of `bins`,
+        # because that is the beginning of the last bin.
+        bins[-2] + y_padding,
+    )
 
     return bins, histograms, x_ranges, y_range
 
 
+@_needs_matplotlib_backend
```
Comment (dakshvar22) [inline rasa/utils/plotting.py:212]: Why was this not needed earlier and needed now?

Comment (JEM-Mosig) [reply rasa/utils/plotting.py:212]: It was there before as decorator for `plot_histogram`.

Comment (JEM-Mosig) [reply rasa/utils/plotting.py:212]: 🤔 Why does github not show this?

**Review by dakshvar22 — APPROVED** [review]
💯

**Pull request merged**